### PR TITLE
Release 0.2

### PR DESCRIPTION
**Pull request by jvatic**
@titanous

### CODE REVIEW

File: lib/tentd/api/authentication_lookup.rb [hunk 1]
```diff
@@ -18,14 +18,14 @@ def action(env)
...
-          ).first
+          ).where("app_authorizations.tent_expires_at IS NULL OR app_authorizations.tent_expires_at > ?", Time.now.to_i).first
         end
         env.potential_auth = Model::Following.first(:user_id => TentD::Model::User.current.id, :mac_key_id => mac_key_id) unless env.potential_auth
         if env.potential_auth
           env.hmac.secret = env.potential_auth.mac_key
           env.hmac.algorithm = env.potential_auth.mac_algorithm
         elsif mac_key_id
-          return [401, {'WWW-Authenticate' => 'MAC'}, ['Invalid MAC Key ID']]
+          return [401, {'WWW-Authenticate' => 'MAC', 'Content-Type' => MEDIA_TYPE}, [{ 'error' => 'Invalid MAC Key ID' }.to_json]]
```
Comment (titanous) [inline lib/tentd/api/authentication_lookup.rb]: How about a helper function that handles error responses?

File: lib/tentd/api/followers.rb [hunk 2]
```diff
@@ -3,8 +3,26 @@ class API
     class Followers
       include Router
 
+      class ParseLookupKey < Middleware
+        def action(env)
+          id_or_entity = env.params.delete(:captures).first
+
+          if id_or_entity =~ /^https?:\/\//
```
Comment (titanous) [inline lib/tentd/api/followers.rb]: Prefer `\A` over `^`.

File: lib/tentd/api/followers.rb [hunk 3]
```diff
@@ -110,8 +129,31 @@ def action(env)
...
       end
 
+      class EntityRedirect < Middleware
+        def action(env)
+          return env unless env.params.has_key?(:follower_entity)
+
+          follower = Model::Follower.select(:id, :public, :public_id).where(:entity => env.params.follower_entity, :user_id => Model::User.current.id).first
+          if follower && !follower.public? && !(env.full_read_authorized || authorize_env?(env, :self))
+            follower = Model::Follower.find_with_permissions(follower.id, env.current_auth)
+            raise Unauthorized unless follower
```
Comment (titanous) [inline lib/tentd/api/followers.rb]: This should actually 404.

File: lib/tentd/api/followers.rb [hunk 4]
```diff
@@ -110,8 +129,31 @@ def action(env)
...
+          end
+
+          unless follower
+            raise Unauthorized unless env.full_read_authorized
```
Comment (titanous) [inline lib/tentd/api/followers.rb]: I think this should just 404 as well.

File: lib/tentd/api/followers.rb [hunk 5]
```diff
@@ -147,6 +189,22 @@ def action(env)
         end
       end
 
+      class CountHeader < Middleware
+        def action(env)
+          count_env = env.dup
+          count_env.params.return_count = true
+          count = GetMany.new(@app).call(count_env)[2][0]
```
Comment (titanous) [inline lib/tentd/api/followers.rb]: This is weird. Also, this header should only be added on HEAD requests.

File: lib/tentd/api/followings.rb [hunk 6]
```diff
@@ -26,6 +44,27 @@ def action(env)
...
+          return env unless env.params.has_key?(:following_entity)
+
+          following = Model::Following.select(:id, :public, :public_id).where(:entity => env.params.following_entity, :user_id => Model::User.current.id).first
+          if following && !following.public && !authorize_env?(env, :read_followings)
+            following = Model::Following.find_with_permissions(following.id, env.current_auth)
+            raise Unauthorized unless following
+          end
+
+          unless following
+            raise Unauthorized unless authorize_env?(env, :read_followings)
```
Comment (titanous) [inline lib/tentd/api/followings.rb]: Ditto on the 404s here.

File: lib/tentd/api/middleware.rb [hunk 7]
```diff
@@ -25,7 +29,24 @@ def call(env)
...
+        uri.scheme = env['rack.url_scheme']
+
+        port = (env.HTTP_X_FORWARDED_PORT || env.SERVER_PORT).to_i
+        uri.port = port unless [80, 443].include?(port)
+
+        uri
+      end
+
+      def serialize_params(params)
+        "#{params.inject([]) { |m, (k,v)| m << "#{k}=#{URI.encode_www_form_component(v)}"; m }.join('&')}"
```
Comment (titanous) [inline lib/tentd/api/middleware.rb]: Why is this wrapped in `"#{}"`?

File: lib/tentd/api/pagination_header.rb [hunk 8]
```diff
@@ -0,0 +1,53 @@
+module TentD
+  class API
+    class PaginationHeader < Middleware
+      def action(env)
+        return env unless env.response.kind_of?(Array) && env.response.size > 0
+
+        env['response.headers'] ||= {}
+        pagination = [%(<#{prev_uri(env).to_s}>; rel="prev"), %(<#{next_uri(env).to_s}>; rel="next")]
+        if env['response.headers']['Link']
+          env['response.headers']['Link'] += ",#{pagination.join(',')}"
```
Comment (titanous) [inline lib/tentd/api/pagination_header.rb]: Add a space after the first `,`.

File: lib/tentd/api/pagination_header.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        else
+          env['response.headers']['Link'] = pagination.join(',')
```
Comment (titanous) [inline lib/tentd/api/pagination_header.rb]: `.join(', ')` here and above.

File: lib/tentd/api/pagination_header.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+        uri.query = serialize_params(build_next_params(env))
+        uri
+      end
+
+      def build_next_params(env)
+        params = env.params.dup
+        resource = env.response.last
+
+        params.before_id = resource.public_id
```
Comment (titanous) [inline lib/tentd/api/pagination_header.rb]: Check if we're on the last page by checking whether the size of the response is the size of the requested page size.

File: lib/tentd/api/followers.rb (hunk 5 above)
Comment (titanous) [inline lib/tentd/api/followers.rb]: Missed that this happens only on head requests. Why does `GetMany` get called twice?

Comment (jvatic) [inline lib/tentd/api/followers.rb]: The first call to `GetMany` is the same as a GET request (sets the Content-Length header), and the second call gets the count (which is not the same as `env.response.size`.)

Comment (titanous) [inline lib/tentd/api/followers.rb]: Ah.

File: lib/tentd/api/middleware.rb [hunk 11]
```diff
@@ -25,7 +29,28 @@ def call(env)
...
+        uri.port = port unless [80, 443].include?(port)
+
+        uri
+      end
+
+      def serialize_params(params)
+        params.inject([]) { |m, (k,v)| m << "#{k}=#{URI.encode_www_form_component(v)}"; m }.join('&')
+      end
+
+      def error_response(status, headers, error)
```
Comment (titanous) [inline lib/tentd/api/middleware.rb]: Perhaps you could put the headers last and default so that they can be omitted unless needed?

Comment (jvatic) [inline lib/tentd/api/middleware.rb]: Yes, I was trying to keep with the status, headers, body convention but I suppose that doesn't really matter in this case.

File: lib/tentd/api/posts.rb [hunk 12]
```diff
@@ -277,24 +325,87 @@ def action(env)
...
+      class GetMentions < Middleware
+        def action(env)
+          return env unless post = env.response
+          env.post = post
+          env.response = post.public_mentions(env.params.slice(:before_id, :since_id, :limit, :post_types, :return_count))
+          env
+        end
+      end
+
+      class MentionsCountHeader < Middleware
```
Comment (titanous) [inline lib/tentd/api/posts.rb]: It seems like these count header middlewares have a lot of boilerplate, is it possible to DRY it up a bit?

File: lib/tentd/api/followers.rb (hunk 4 above)
Comment (jvatic) [inline lib/tentd/api/followers.rb]: Then [`GetOne`](https://github.com/tent/tentd/blob/9aee6cfca3ae743efcf5e000b31ad6aee8fa2bd5/lib/tentd/api/followers.rb#L166) should 404 also?

Comment (titanous) [inline lib/tentd/api/followers.rb]: Yes.

File: db/migrations/003_add_tent_expires_at_to_app_authorizations.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+
+Sequel.migration do
+  change do
+    alter_table(:app_authorizations) do
+      add_column :tent_expires_at, Integer
```
Comment (titanous) [inline db/migrations/003_add_tent_expires_at_to_app_authorizations.rb]: Can we just call this column `expires_at`?

File: lib/tentd/api/pagination_header.rb (hunk 10 above)
Comment (jvatic) [inline lib/tentd/api/pagination_header.rb]: I also missed the case when `before_id` and `since_id` have reversed roles when ordering `asc` vs `desc`.

**Pull request closed without merging**